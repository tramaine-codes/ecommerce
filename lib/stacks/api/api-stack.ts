import * as cdk from 'aws-cdk-lib';
import type * as cognito from 'aws-cdk-lib/aws-cognito';
import type * as dynamo from 'aws-cdk-lib/aws-dynamodb';
import type { Construct } from 'constructs';
import type { Config } from '../../infrastructure/config/config.js';
import { RestApi } from './api-gateway/rest-api.js';

interface ApiStackProps extends cdk.StackProps {
  readonly cognitoUserPools: cognito.IUserPool[];
  readonly config: Config;
  readonly productsTable: dynamo.ITableV2;
}

export class ApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id, props);

    const { cognitoUserPools, config, productsTable } = props;

    new RestApi(this, {
      cognitoUserPools,
      config,
      productsTable,
    });
  }
}
