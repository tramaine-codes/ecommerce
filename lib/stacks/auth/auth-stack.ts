import * as cdk from 'aws-cdk-lib';
import type * as cognito from 'aws-cdk-lib/aws-cognito';
import type { Construct } from 'constructs';
import type { Config } from '../../infrastructure/config/config.js';
import { ApiUserPool } from './pools/api-user-pool.js';

interface AuthStackProps extends cdk.StackProps {
  readonly config: Config;
}

export class AuthStack extends cdk.Stack {
  readonly apiUserPool: cognito.IUserPool;

  constructor(scope: Construct, id: string, props: AuthStackProps) {
    super(scope, id, props);

    this.apiUserPool = new ApiUserPool(this, props);
  }
}
