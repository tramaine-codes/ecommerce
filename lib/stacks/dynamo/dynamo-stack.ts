import * as cdk from 'aws-cdk-lib';
import type * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import type { Construct } from 'constructs';
import { ProductsResources } from './products/products-resources.js';

export class DynamoStack extends cdk.Stack {
  readonly productsTable: dynamodb.ITableV2;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const { productsTable } = new ProductsResources(this);

    this.productsTable = productsTable;
  }
}
