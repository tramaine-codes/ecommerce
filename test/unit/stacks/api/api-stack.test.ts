import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { expect, test } from 'vitest';
import { Config } from '../../../../lib/infrastructure/config/config.js';
import { ApiStack } from '../../../../lib/stacks/api/api-stack.js';
import { AuthStack } from '../../../../lib/stacks/auth/auth-stack.js';
import { DynamoStack } from '../../../../lib/stacks/dynamo/dynamo-stack.js';

const app = new cdk.App();
const config = new Config();
const { productsTable } = new DynamoStack(app, 'DynamoStack');
const { apiUserPool } = new AuthStack(app, 'AuthStack', {
  config,
});
const stack = new ApiStack(app, 'ApiStack', {
  cognitoUserPools: [apiUserPool],
  config,
  productsTable,
});

const template = Template.fromStack(stack);

test('matches the snapshot', () => {
  expect(template.toJSON()).toMatchSnapshot();
});
