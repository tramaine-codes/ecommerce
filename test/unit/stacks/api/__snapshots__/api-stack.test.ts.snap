// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "Outputs": {
    "RestApiEndpoint0551178A": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "RestApi0C43BF4B",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "RestApiDeploymentStageprod3855DE66",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "RestApi0C43BF4B": {
      "Properties": {
        "Name": "RestApi",
        "Tags": [
          {
            "Key": "Name",
            "Value": "EcommerceApi",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "RestApiAccount7C83CF5A": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "RestApi0C43BF4B",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "RestApiCloudWatchRoleE3ED6605",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "RestApiCloudWatchRoleE3ED6605": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "EcommerceApi",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "RestApiCognitoAuthorizerD2BF1AF9": {
      "Properties": {
        "IdentitySource": "method.request.header.Authorization",
        "Name": "ApiStackRestApiCognitoAuthorizerA72935B7",
        "ProviderARNs": [
          {
            "Fn::ImportValue": "AuthStack:ExportsOutputFnGetAttEcommerceApiPool917914B5Arn0423AD07",
          },
        ],
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
        "Type": "COGNITO_USER_POOLS",
      },
      "Type": "AWS::ApiGateway::Authorizer",
    },
    "RestApiDeployment180EC5031ef61031a993eeb3a63497b09b60e04a": {
      "DependsOn": [
        "RestApiproductsproductIdGET7F758B01",
        "RestApiproductsproductId1A378357",
        "RestApiproductsPOST6AB87446",
        "RestApiproducts287512B0",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "RestApiDeploymentStageprod3855DE66": {
      "DependsOn": [
        "RestApiAccount7C83CF5A",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "RestApiDeployment180EC5031ef61031a993eeb3a63497b09b60e04a",
        },
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
        "StageName": "prod",
        "Tags": [
          {
            "Key": "Name",
            "Value": "EcommerceApi",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "RestApiProductsResourcesApiHandler0BDBD24C": {
      "DependsOn": [
        "RestApiProductsResourcesApiHandlerServiceRoleDefaultPolicyB07938D5",
        "RestApiProductsResourcesApiHandlerServiceRole9BE652BD",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "60ee5e75c45f40fcc59e086e599d531c63ac04bea7e3a2d9d1177dabf9ff6f50.zip",
        },
        "Environment": {
          "Variables": {
            "PRODUCTS_TABLE_NAME": {
              "Fn::ImportValue": "DynamoStack:ExportsOutputRefProductsResourcesProductsTable5E5C544881DF1299",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "RestApiProductsResourcesApiHandlerServiceRole9BE652BD",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "Name",
            "Value": "EcommerceApi",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
    },
    "RestApiProductsResourcesApiHandlerServiceRole9BE652BD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "EcommerceApi",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RestApiProductsResourcesApiHandlerServiceRoleDefaultPolicyB07938D5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "DynamoStack:ExportsOutputFnGetAttProductsResourcesProductsTable5E5C5448Arn76B6C8C1",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "DynamoStack:ExportsOutputFnGetAttProductsResourcesProductsTable5E5C5448Arn76B6C8C1",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RestApiProductsResourcesApiHandlerServiceRoleDefaultPolicyB07938D5",
        "Roles": [
          {
            "Ref": "RestApiProductsResourcesApiHandlerServiceRole9BE652BD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RestApiproducts287512B0": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "RestApi0C43BF4B",
            "RootResourceId",
          ],
        },
        "PathPart": "products",
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiproductsPOST6AB87446": {
      "Properties": {
        "AuthorizationScopes": [
          "EcommerceApiResourceService/ecommerceapi.write",
        ],
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": {
          "Ref": "RestApiCognitoAuthorizerD2BF1AF9",
        },
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "RestApiProductsResourcesApiHandler0BDBD24C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiproducts287512B0",
        },
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiproductsPOSTApiPermissionApiStackRestApiF34D39F9POSTproductsA5079520": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "RestApiProductsResourcesApiHandler0BDBD24C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              {
                "Ref": "RestApiDeploymentStageprod3855DE66",
              },
              "/POST/products",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiproductsPOSTApiPermissionTestApiStackRestApiF34D39F9POSTproductsFABF85C3": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "RestApiProductsResourcesApiHandler0BDBD24C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/POST/products",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiproductsproductId1A378357": {
      "Properties": {
        "ParentId": {
          "Ref": "RestApiproducts287512B0",
        },
        "PathPart": "{productId}",
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiproductsproductIdGET7F758B01": {
      "Properties": {
        "AuthorizationScopes": [
          "EcommerceApiResourceService/ecommerceapi.read",
        ],
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": {
          "Ref": "RestApiCognitoAuthorizerD2BF1AF9",
        },
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "RestApiProductsResourcesApiHandler0BDBD24C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiproductsproductId1A378357",
        },
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiproductsproductIdGETApiPermissionApiStackRestApiF34D39F9GETproductsproductIdA71CE43A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "RestApiProductsResourcesApiHandler0BDBD24C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              {
                "Ref": "RestApiDeploymentStageprod3855DE66",
              },
              "/GET/products/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiproductsproductIdGETApiPermissionTestApiStackRestApiF34D39F9GETproductsproductId35160203": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "RestApiProductsResourcesApiHandler0BDBD24C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/GET/products/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
